eclipse/che
# Che CLI Improvements

### What does this PR do?
1. Adds in error message to check for missing "-it" on the command line.

2. Adds in better, friendlier help when missing different volume mounts.

3. Changes the ":/che" mount to be ":/data" to be consistent for different assemblies.  Replaced a hard coded "/che" reference in the docker-compose template, and instead passed in the CHE_CONTAINER_ROOT into puppet to be used.

4. Adds a new cleanup() method to catch any exit scenario. If the product exits for any reason with return code "3", we will print out a message to the user to tell them how they can rerun to get the cli.log.

5. If in development mode, adds a check to pull the CHE_DEBUG_PORT from the user's che.env file and then performs a port check on that during the pre-flight checks to verify that it is available.  We will need to create a matching improvement for Codenvy assembly to solve this issue.
https://github.com/codenvy/codenvy/issues/1282

6. Improved the usage output format. Also moved usage() so that it can appear before any docker checks.

7. Removed the need to have ':/cli' volume mount. We now just write the cli.log every time to the ':/data' folder.

8. Adds a new check that if the system is not in --offline mode, then verifies that the DNS resolver is working to dockerhub.com. If not, prints out an error message with users on how to fix it with their Docker settings.

To test this version of the CLI:
1. git clone http://github.com/eclipse/che
2. git checkout cli-improvements
3. cd dockerfiles/base;  bash build.sh
4. cd dockerfiles/cli; bash build.sh

The syntax has slightly changed, so please check the /docs/README.md in this branch for the updates.
